Mod support

I-Have-No-Idea-What-IAmDoing opened the pull request:
Not Really Fixes But Improves??? #584 

I made a mod loader with a priority/load order 
![image](https://user-images.githubusercontent.com/55067948/89960067-89046280-dc03-11ea-9d55-67569c4c4d12.png)

it loads pck files which has it own caveats like:
- 2gb size limit (this probably won't be a problem but might be with giant mods)
- one file can only be changed by one mod (so 2 mods can't edit the main menu, that where the load order comes in)
- no way to unload mods
- mods are not persistent (any changes made by mods are lost when the game is closed, this is either a good or bad thing)
- they are harder to make than just editing files

you should be able to load c# code by loading the dll file but i have no idea how to test it, so mods should use gdscript for scripting. i am also unsure if you can modify the base game code since it compiled into the dll file.

mods are detected by a mod_info.json file which holds info like the name, author, version, a description of the mod, and the location of a dll file (if it needed). the actual mod file is the mod.pck file. each mod has it own folder but the name of that folder does not matter.

there 3 problems which i don't know how to fix which is why this is a draft right now:
- When replacing existing asset, you have to change scene for this to take effect(go into the game and go back) the loader should already reload the main menu but the changes still don't take effect until you change scene
(Edit: This is not true, the asset only changes when you load a save... weird)
- Json files are not able to be changed since they are readonly and are assigned when the game is starting (this prevent mods that add new music,or add new organelles, or anything in the simulation parameters)
- i unsure on how you change/add c# file, i know you load a dll but i am unsure, how to do it, if that possible or even if my implementation actually works right now 

there also some changes that i want to make like:
- a reset button (since you can't unload mods this will just load the pck file of the release, this will reset any file but it won't get rid of any new ones) ✅
- a way to easily move the load order beside unloading and reloading (either with a drag and drop or just buttons) ✅

it should be simple to disable this feature as this is fairly self-contained, you just have to disable the button that opens the mod loader

Here is 4 mods that i made that you can use to test with:
https://we.tl/t-yafn4CjqIi

- Mod Template - This is a template for modders to use , it only contain the mod_info.json
- Alternate Main Menu - This mod show adding new assets and changing scene layout
- Thriv 2 (Better Asset Flip) - This shows changing existing assets but this is the one with the bug
- Quick Reset - This show scripting (including calling c# methods) and sound (This one has the most compatibility with other mods)

all of them can be seen and tested in the main menu

hhyyrylainen commented:
This is a pretty good start. Glad to see someone working on this.

Some comments regarding the problems:
- I think that the best way would be to store a list of mods to load in Settings and making a auto load mod loader that would load them early in the startup. This way they can modify even the json files and restarting the game is possible, though a game restart would be required to unload mods or load json modifying mods. This probably also helps with the need to change the scene.
- If there is a 2GB pck size limit thrive itself will hit that probably at some point, hopefully it isn't a problem and Godot devs will fix it.
- Making proper hooks / mod framework can be left for later to allow multiple mods to modify the same kind of things
- I think that C# code can use reflection to modify existing loaded code. At least if we add hook points into our code it should be possible for mods to register running code that way.

I-Have-No-Idea-What-IAmDoing commented:
I just added the auto-loading which is determined by the mod_info.json 
I also added buttons for resetting the game and for moving the mod order which should be handy 

but the test mods i have made are currently incompatible so i need to remake them

I-Have-No-Idea-What-IAmDoing commented:
I can't figure it out, i already spent a week just trying to fix this problem and i think it's too advance for me and beside i already implemented a workaround (use the autoloader). So i guess it ready for review
________________

Stuff I Found Out For Future Reference:
* Loading The Game Reloads Icons,Misc Images, Backgrounds But Not Textures (Like The Nucleus), I am unsure about other types of assets
* Returning to the main menu reloads only Misc Images, And Icons 
* Creating new nodes using the replaced asset still uses the old version of that asset
* Changing to a new scene seems to not fix it but i might have just messed up
* It does not seem to be a problem with loading the mod as it seem to replace the asset fine
* Something about loading the game actually like refreshes the assets and not anything else

hhyyrylainen commented:
I think that Godot keeping existing assets loaded might be the reason behind all of that you explain.
It'd be probably good to talk to other people who have tried to mod or add modding to a Godot based game.

Before this can be considered for merging the cleanup and format_script problems need to be fixed.

I-Have-No-Idea-What-IAmDoing commented:
Is there anything else i need to do?

hhyyrylainen commented:
What's the status of the functionality for this? Is this ready for me to test and code review?

I-Have-No-Idea-What-IAmDoing commented:
It should be ready to test

Here a link to some mods you can use to test:
https://we.tl/t-uXDf8MqAzu

The rainbow thrive mod is a autoloaded mod so that one should be installed separately

hhyyrylainen commented:
I didn't check the code yet, but I have a few comments about the sample mods.
It seems like most of the mods contain all of the thrive files, even ones that weren't modified? Isn't this highly wasteful in terms of download size and mods overriding other mods with the default assets?
Also it'd be nice to have an example mod that used C# to test if loading code like that works.

It seems that the autoload property is not included in the mod description, isn't that a limitation? Shouldn't mods define *when* to load them (autoload) or after game initialization?

hhyyrylainen commented on .gitignore:
```diff
@@ -133,3 +133,4 @@ package-lock.json
 /dis
 /ena
 /fast_build_mode.json
+mods/
```
This should be /mods to just ignore a mods folder in the base directory.

hhyyrylainen commented on Thrive.csproj:
```diff
@@ -41,6 +41,7 @@
     <ErrorReport>prompt</ErrorReport>
     <WarningLevel>4</WarningLevel>
     <ConsolePause>false</ConsolePause>
+    <LangVersion>latest</LangVersion>
```
Is it right that this is added to just this group? Also mono and whatever build tool is used might have different C# versions, so shouldn't an explicit version be used instead?

I-Have-No-Idea-What-IAmDoing replied:
I am not sure, that just how it worked for me

hhyyrylainen replied:
I mean this might not work for everyone. So I'd prefer the explicit version setting in every target, or this line to be removed.

I-Have-No-Idea-What-IAmDoing replied:
I think it best for me to just remove it

hhyyrylainen commented on Thrive.csproj:
```diff
@@ -65,6 +66,9 @@
     <Compile Include="src\engine\SceneManager.cs" />
     <Compile Include="src\engine\ScreenShotTaker.cs" />
     <Compile Include="src\engine\StartupActions.cs" />
+    <Compile Include="src\general\AutoModLoader.cs" />
+    <Compile Include="src\general\ModInfo.cs" />
+    <Compile Include="src\general\ModLoader.cs" />
```
I think these should be in the engine folder instead of general.

hhyyrylainen commented on Thrive.csproj:
```diff
@@ -234,9 +238,6 @@
   </ItemGroup>
   <ItemGroup>
     <Folder Include="shaders\" />
-    <Folder Include="src\microbe_stage\organelle_components\" />
-    <Folder Include="src\auto-evo\steps\" />
-    <Folder Include="src\auto-evo\simulation\" />
```
These entries shouldn't be deleted.

hhyyrylainen commented on Thrive.csproj.old:
```diff
@@ -0,0 +1,283 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Remove this backup file from being included in this PR.

hhyyrylainen commented on Thrive.sln.old:
```diff
@@ -0,0 +1,49 @@
+Microsoft Visual Studio Solution File, Format Version 12.00
```
Another backup file.

hhyyrylainen commented on src/general/AutoModLoader.cs:
```diff
@@ -0,0 +1,65 @@
+using Directory = System.IO.Directory;
+using File = System.IO.File;
+
+public class AutoModLoader : Node
```
This should have a comment describing what this does.

hhyyrylainen commented on src/general/AutoModLoader.cs:
```diff
@@ -0,0 +1,65 @@
+{
+    private AutoModLoader()
+    {
+        DirectoryInfo modFolder = Directory.CreateDirectory($"{Directory.GetCurrentDirectory()}\\mods");
```
Mods should really be in the Godot "user://" directory as the game could be installed in a read only (or administrator writable) folder. So mods need to be elsewhere than in the game folder.

hhyyrylainen commented on src/general/AutoModLoader.cs:
```diff
@@ -0,0 +1,65 @@
+    private AutoModLoader()
+    {
+        DirectoryInfo modFolder = Directory.CreateDirectory($"{Directory.GetCurrentDirectory()}\\mods");
+        foreach (DirectoryInfo currentMod in modFolder.EnumerateDirectories())
```
The previous comment likely requires this to be changed to use the Godot engine provided directory iterator.

hhyyrylainen commented on src/general/AutoModLoader.cs:
```diff
@@ -0,0 +1,65 @@
+                JsonConvert.DeserializeObject<ModInfo>(ReadJSONFile(currentMod.FullName + "/mod_info.json"));
+            if (currentModInfo.AutoLoad)
+            {
+                if (string.IsNullOrEmpty(currentModInfo.Dll))
+                {
+                    if (File.Exists(currentMod.FullName + "/" + currentModInfo.Dll))
+                    {
+                        Assembly.LoadFile(currentMod.FullName + "/" + currentModInfo.Dll);
+                    }
+                }
+
+                if (!File.Exists(currentMod.FullName + "/mod.pck"))
+                {
+                    GD.Print("Fail to find mod file: " + currentModInfo.ModName);
+                    continue;
+                }
+
+                if (ProjectSettings.LoadResourcePack(currentMod.FullName + "/mod.pck", true))
+                {
+                    GD.Print("Loaded mod: " + currentModInfo.ModName);
+                }
+                else
+                {
+                    GD.Print("Failed to load mod: " + currentModInfo.ModName);
+                }
```
This is duplicated logic with the other mod load class.

hhyyrylainen commented on src/general/AutoModLoader.cs:
```diff
@@ -0,0 +1,65 @@
+                {
+                    if (File.Exists(currentMod.FullName + "/" + currentModInfo.Dll))
+                    {
+                        Assembly.LoadFile(currentMod.FullName + "/" + currentModInfo.Dll);
```
Allowing to load the assembly from within a pck would be nice, similarly to how the main game dll is included in Thrive.pck

I-Have-No-Idea-What-IAmDoing replied:
i am not sure how to really do that as LoadFile uses a path and you can't use a godot path with it

hhyyrylainen replied:
It seems that Assembly.Load (https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly.load?view=netcore-3.1) might work if the raw data is read with godot and then passed there.
Another option would be to ask Godot to translate a godot path to an absolute path and then proceed.

hhyyrylainen commented on src/general/AutoModLoader.cs:
```diff
@@ -0,0 +1,65 @@
+                }
+                else
+                {
+                    GD.Print("Failed to load mod: " + currentModInfo.ModName);
```
After the game starts it'd be nice to show a popup if some mods failed to load.

hhyyrylainen commented on src/general/AutoModLoader.cs:
```diff
@@ -0,0 +1,65 @@
+
+            var currentModInfo =
+                JsonConvert.DeserializeObject<ModInfo>(ReadJSONFile(currentMod.FullName + "/mod_info.json"));
+            if (currentModInfo.AutoLoad)
```
There should be some way to specify which mods to load other than by just moving or deleting the files.

hhyyrylainen commented on src/general/AutoModLoader.cs:
```diff
@@ -0,0 +1,65 @@
+        QueueFree();
+    }
+
+    private static string ReadJSONFile(string path)
```
Duplicated code

hhyyrylainen commented on src/general/ModInfo.cs:
```diff
@@ -0,0 +1,16 @@
+public class ModInfo : Node
+{
+    [JsonProperty("Name")]
+    public string ModName;
```
Why does this property have different name in the json? Why not just have `public string Name` ?

I-Have-No-Idea-What-IAmDoing replied:
Calling it Name gave me a warning so i just changed it to get rid of it

hhyyrylainen replied:
I see that this inherits from Node. There doesn't seem a reason for that. It will probably work if you don't inherit this class from anything.

hhyyrylainen commented on src/general/ModInfo.cs:
```diff
@@ -0,0 +1,16 @@
+using Godot;
+using Newtonsoft.Json;
+
+public class ModInfo : Node
```
Class should have a doc comment.

hhyyrylainen commented on src/general/ModLoader.cs:
```diff
@@ -0,0 +1,320 @@
+using Directory = System.IO.Directory;
+using File = System.IO.File;
+
+public class ModLoader : Control
```
Should be documented as well.

And it'd be nice to split the logic and UI, so that there is a ModLoader (perhaps an autoload, combined with the autoload mod loader), and a separate ModManagerUI that inherits Control.

hhyyrylainen commented on src/general/ModLoader.cs:
```diff
@@ -0,0 +1,320 @@
+
+        confirmationPopup = GetNode<ConfirmationDialog>(ConfirmationPopupPath);
+
+        DirectoryInfo modFolder = Directory.CreateDirectory($"{Directory.GetCurrentDirectory()}\\mods");
```
Duplicated code. Should probably be made into a method for easily refreshing the list of available mods.

hhyyrylainen commented on src/general/ModLoader.cs:
```diff
@@ -0,0 +1,320 @@
+    {
+        var tempModInfo = modList[index];
+        modInfoName.Text = tempModInfo.ModName;
+        modInfoAuthor.Text = "Author: " + tempModInfo.Author;
+        modInfoVersion.Text = "Version: " + tempModInfo.Version;
```
Splitting these into separate controls would be preferrable to allow for translations. The author value loaded from the mod should be in a separate label by itself.

hhyyrylainen commented on src/general/ModLoader.cs:
```diff
@@ -0,0 +1,320 @@
+    private void OnLoadedModSelected(int index)
+    {
+        var tempModInfo = loadedModList[index];
+        modInfoName.Text = tempModInfo.ModName;
+        modInfoAuthor.Text = "Author: " + tempModInfo.Author;
+        modInfoVersion.Text = "Version: " + tempModInfo.Version;
+        modInfoDescription.Text = tempModInfo.Description;
+        modInfoDescription.Text = tempModInfo.Description;
```
This code looks very similar to what's in OnUnloadedModSelected, is it possible to generalize this to avoid duplication?

hhyyrylainen commented on src/general/ModLoader.cs:
```diff
@@ -0,0 +1,320 @@
+            return;
+        }
+
+        var selectedItem = unloadedItemList.GetSelectedItems()[0];
```
What if there are multiple selected items?

I-Have-No-Idea-What-IAmDoing replied:
You are only able to select a single item as the item list select mode is set to be single

hhyyrylainen replied:
Okay.

hhyyrylainen commented on src/general/ModLoader.cs:
```diff
@@ -0,0 +1,320 @@
+        loadedModList.Add(modList[selectedItem]);
+        loadedItemList.AddItem(modList[selectedItem].ModName);
+        modList.RemoveAt(selectedItem);
+        unloadedItemList.RemoveItem(selectedItem);
```
You should add some extra blank lines here (and in other places where there are 4-5 or more lines without blanks in a row.

hhyyrylainen commented on src/general/ModLoader.cs:
```diff
@@ -0,0 +1,320 @@
+        loadedModList.Clear();
+        unloadedItemList.Clear();
+        loadedItemList.Clear();
+        DirectoryInfo modFolder = Directory.CreateDirectory($"{Directory.GetCurrentDirectory()}\\mods");
```
Code duplication.

hhyyrylainen commented on src/general/ModLoader.cs:
```diff
@@ -0,0 +1,320 @@
+        }
+    }
+
+    private void OnLoadPressed()
```
What happens if user presses this multiple times? Does this work? Do DLL containing mods cause issues? What if we ever add callbacks for mods to react to loading?
I think this needs some changes to not load already loaded mods.

hhyyrylainen commented on src/general/ModLoader.cs:
```diff
@@ -0,0 +1,320 @@
+        }
+
+        GD.Print("All mods loaded");
+        SceneManager.Instance.ReturnToMenu();
```
I'm a bit suspicious why this is done and not a normal return to the menu? Does it has to do with mods otherwise not being able to overwrite things properly?

I-Have-No-Idea-What-IAmDoing replied:
What is the normal way to return to the main menu?

hhyyrylainen replied:
Trigger the return callback like with the back button there is in the mod manager.
Also the same thing in the options menu and save load menu.

I-Have-No-Idea-What-IAmDoing replied:
Yeah i do that as if i don't mods don't load properly

hhyyrylainen commented on src/general/ModLoader.cs:
```diff
@@ -0,0 +1,320 @@
+    private void ResetGame()
+    {
+        GUICommon.Instance.PlayButtonPressSound();
+        if (!File.Exists(Directory.GetCurrentDirectory() + "/Thrive.pck"))
```
Would it be possible to ask godot what pck we have loaded? As the path might not always work. AFAIK you can rename the godot game executable and pck file and it'll still work. This would fail with that.

I-Have-No-Idea-What-IAmDoing replied:
It does not seem to be possible to do that but i could ask the player to select a file to reset from instead

hhyyrylainen replied:
Another idea for reset, is to just require the game to be restarted. And before that marking all mods not to be loaded.

I-Have-No-Idea-What-IAmDoing replied:
I could do that

hhyyrylainen commented on src/general/ModLoader.cs:
```diff
@@ -0,0 +1,320 @@
+        GUICommon.Instance.PlayButtonPressSound();
+        if (!File.Exists(Directory.GetCurrentDirectory() + "/Thrive.pck"))
+        {
+            GD.Print("Fail to find Thrive");
```
This should show an error popup.

hhyyrylainen commented on src/general/ModLoader.cs:
```diff
@@ -0,0 +1,320 @@
+        }
+        else
+        {
+            GD.Print("Reset failed");
```
Error popup here as well would be good.

hhyyrylainen commented on src/general/ModLoader.cs:
```diff
@@ -0,0 +1,320 @@
+                return;
+            }
+
+            loadedItemList.MoveItem(selectedItemIndex, selectedItemIndex - 1);
+            selectedItemIndex = loadedItemList.GetSelectedItems()[0];
+            MoveItem(loadedModList, selectedItemIndex + 1, selectedItemIndex);
```
Would it be possible to just run the MoveItems in all cases? It seems that some godot child reordering code just calls the move to put things in the right order even if they already are.
This would reduce a lot of similar code.

hhyyrylainen commented on src/general/ModLoader.cs:
```diff
@@ -0,0 +1,320 @@
+        }
+    }
+
+    private void MoveItem(List<ModInfo> list, int oldIndex, int newIndex)
```
Is there no inbuilt method for this?
What about using a LINQ sort on the list to get it to the wanted order?

hhyyrylainen reviewed (changes requested):
Quite a few comments on the code as this is the first time me checking it.
One more general note: the list of mods to load should be saved in Settings so that on each game start you don't have to reapply all the mods.

I-Have-No-Idea-What-IAmDoing commented:
I have just implemented all the changes, going to create mods for testing later

hhyyrylainen commented:
You should merge master into this branch as there's been quite a few changes there (especially related to Settings), that need fixing before this can be merged into master.

I-Have-No-Idea-What-IAmDoing commented:
Ok i just merged it but right now i am having issues with the mods being stored in settings and it got me thinking, should the mods be stored in a different file as it feel like it fits right in there.

hhyyrylainen commented:
The mod info could be stored in a separate file, but then it would need a lot of the same logic as Settings. So in my opinion having "active mods settings" in Settings is quite logical.

I-Have-No-Idea-What-IAmDoing commented:
When running the formatting check, if i don't have localization in my scene does it give a error when checking localization?

hhyyrylainen commented:
It doesn't, but by default texts and some other properties of Noders are considered translateable. If you haven't updated the localization files to have new texts in them, formatting check will complain.

I-Have-No-Idea-What-IAmDoing commented:
Whenever i try to run the localization checks on my machine it keeps on failing, i keep getting this error:

**ERROR: Failed to run translation generation to check if current files are up to date**

here are the logs if they help:
https://pastebin.com/yeNnyL46

hhyyrylainen commented:
It looks like translations check in CI is now working. Do you have pybabel installed properly locally? It's a bit weird that it doesn't print anything. You should try manually running the pybabel extract command to see if it does anything (remember to apply quotes to the path names with spaces).

I-Have-No-Idea-What-IAmDoing commented:
Yeah i was using a old version of python, so i had to reinstall it for it to work

hhyyrylainen commented:
Is this ready to review now? Before merging, I'd like to verify that code containing mods can be loaded correctly, because once we officially include mod support in the game, the mod format needs to be relatively stable so that we won't have to unnecessarily break existing mods.

I-Have-No-Idea-What-IAmDoing commented:
It should be, it been awhile since i worked on this since i was busy, so i don't have any mods to test with yet
also i don't think mod compatibility would be a concern yet as mods are incompatible between versions until we have a modding api

hhyyrylainen commented:
There's a good chance that asset replacement mods and code mods, that use reflection to find something specific to change, have a chance to work between different versions of the game.

I-Have-No-Idea-What-IAmDoing commented:
So i been testing and working on this for a while and i don't think i can make the c# mods work and as i don't really know what i am really doing here for this and i even tried asking on the godot discord and it seems c# modable projects are not really common, so i don't know if it is possible to do this well

the problem i am running into right now is that i am stuck on loading and getting the dll file as if i try to load it right now godot crashes (without errors???) and i am unsure why right now or how to fix it and actually getting the dll file is tricky (right now the mod make have to specify the path of the dll file) as it need to be loaded before the pck file (at least that what the doc say) but the dll is included inside of it but you can try to do it externally which is what i am doing but the path is weird with it. 

So right now my Godot is borked and so i can't even open the project and also i am running out of free time to commit on working on this project so i hope someone else can take on this and try to finish what i started, i still try to work on this when i can but i think i have to take a break for this one right now

I-Have-No-Idea-What-IAmDoing commented:
This is replaced by this PR: #2381
